parallel build with multitarget : process cannot access the file

### Reqnroll Version

2.0.3

### Which test runner are you using?

NUnit

### Test Runner Version Number

4.1.0

### .NET Implementation

.NET 6.0

### Test Execution Method

Azure DevOps Pipeline Task – PLEASE SPECIFY THE NAME OF THE TASK

### Content of reqnroll.json configuration file

no json file

### Issue Description

Hello, I'm trying to migrate specflow 3.9.74 to reqnroll 2.0.3, because we had an issue to build/test on azure pipelines.
We are using azure task "dotnet build" / "dotnet test" and always get this kind of error :
`##[error]C:\Users\Administrator\.nuget\packages\reqnroll.tools.msbuild.generation\2.0.3\build\Reqnroll.Tools.MsBuild.Generation.targets(93,5): Error : [Reqnroll] System.IO.IOException: The process cannot access the file 'C:\xxxxx.feature.cs' because it is being used by another process.`

Our projects target `net6.0;net48`, and it seems it tries to create .feature.cs files in parallel for each target. I tested with Azure task "vstest" and it is the same behavior. Very rarely, it can also occurs in VS2022.

### Steps to Reproduce

- have a big project with ~500 features files
- build project with multiple targets
- errors of file being written in parallel

### Link to Repro Project

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@
 ## Bug fixes:
 
 * Fix: Reqnroll.Autofac: Objects registered in the global container cannot be relsolved in BeforeTestRun/AfterTestRun hooks (#183)
+* Fix: Process cannot access the file when building a multi-target project (#197)
 
 *Contributors of this release (in alphabetical order):* @ajeckmans, @cimnine, @gasparnagy, @obligaron, @stbychkov
 

--- a/Reqnroll.Tools.MsBuild.Generation/CodeBehindWriter.cs
+++ b/Reqnroll.Tools.MsBuild.Generation/CodeBehindWriter.cs
@@ -32,7 +32,7 @@ namespace Reqnroll.Tools.MsBuild.Generation
             {
                 if (!FileSystemHelper.FileCompareContent(outputPath, testFileGeneratorResult.GeneratedTestCode))
                 {
-                    File.WriteAllText(outputPath, testFileGeneratorResult.GeneratedTestCode, Encoding.UTF8);
+                    WriteAllTextWithRetry(outputPath, testFileGeneratorResult.GeneratedTestCode, Encoding.UTF8);
                 }
             }
             else
@@ -42,10 +42,36 @@ namespace Reqnroll.Tools.MsBuild.Generation
                     Directory.CreateDirectory(directoryPath);
                 }
 
-                File.WriteAllText(outputPath, testFileGeneratorResult.GeneratedTestCode, Encoding.UTF8);
+                WriteAllTextWithRetry(outputPath, testFileGeneratorResult.GeneratedTestCode, Encoding.UTF8);
             }
 
             return outputPath;
+        }
+
+        /// <summary>
+        /// When building a multi-targeted project, the build system may try to write the same file multiple times,
+        /// and this can cause an IOException ("The process cannot access the file because it is being used by another process.").
+        /// See https://github.com/reqnroll/Reqnroll/issues/197
+        /// Once we move to Roslyn-based generation, this problem will go away, but for now, we use a workaround of
+        /// retrying the write operation a few times (the content is anyway the same).
+        /// </summary>
+        private void WriteAllTextWithRetry(string path, string contents, Encoding encoding)
+        {
+            const int maxAttempts = 5;
+            for (int i = 1; i <= maxAttempts; i++)
+            {
+                try
+                {
+                    File.WriteAllText(path, contents, encoding);
+                    return;
+                }
+                catch (IOException)
+                {
+                    if (i == maxAttempts)
+                        throw;
+                    System.Threading.Thread.Sleep(i * 50);
+                }
+            }
         }
     }
 }

--- a/Reqnroll.Utils/FileSystemHelper.cs
+++ b/Reqnroll.Utils/FileSystemHelper.cs
@@ -103,10 +103,16 @@ namespace Reqnroll.Utils
         // files are not the same.
         public static bool FileCompareContent(string filePath1, string fileContent)
         {
-            var currentFileContent = File.ReadAllText(filePath1);
+            try
+            {
+                var currentFileContent = File.ReadAllText(filePath1);
 
-            return string.CompareOrdinal(currentFileContent, fileContent) == 0;
-
+                return string.CompareOrdinal(currentFileContent, fileContent) == 0;
+            }
+            catch (IOException)
+            {
+                return false;
+            }
         }
 
         public static string NormalizeDirectorySeparators(string path)

